core/src/main/java/org/apache/iceberg/actions/rewrite/BinPackStrategy.java

```diff
@@ -0,0 +1,148 @@
+import org.apache.iceberg.util.BinPacking;
+import org.apache.iceberg.util.BinPacking.ListPacker;
+import org.apache.iceberg.util.PropertyUtil;
+
+abstract class BinPackStrategy implements RewriteStrategy {
+
+  /**
+   * Minimum number of files that need to be in a file group to be considered
+   * for rewriting.
+   */
+  public static final String MIN_NUM_INPUT_FILES = "min-num-input-files";
+  public static final int MIN_NUM_INPUT_FILES_DEFAULT = 1;
+  private int minNumInputFiles;
+
+  /**
+   * Minimum number of files we want to be created by file group when being
+   * rewritten.
+   */
+  public static final String MIN_NUM_OUTPUT_FILES = "min-num-output-file";
+  public static final int MIN_NUM_OUTPUT_FILES_DEFAULT = 1;
+  private int minNumOutputFiles;
+
+  /**
+   * Adjusts files which will be considered for rewriting. Files smaller than
+   * MIN_FILE_SIZE_BYTES will be considered for rewriting.
+   * <p>
+   * Defaults to 75% of the target file size
+   */
+  public static final String MIN_FILE_SIZE_BYTES = "min-file-size-bytes";
+  public static final double MIN_FILE_SIZE_DEFAULT_RATIO = 0.75d;
+  private long minFileSize;
+
+  /**
+   * Adjusts files which will be considered for rewriting. Files larger than
+   * MAX_FILE_SIZE_BYTES will be considered for rewriting.
+   * <p>
+   * Defaults to 180% of the target file size
+   */
+  public static final String MAX_FILE_SIZE_BYTES = "max-file-size-bytes";
+  public static final double MAX_FILE_SIZE_DEFAULT_RATIO = 1.80d;
+  private long maxFileSize;
+
+  private long targetFileSize;
+
+  private long maxGroupSize;
+
+  @Override
+  public String name() {
+    return Strategy.BINPACK.name();
+  }
+
+  @Override
+  public Set<String> validOptions() {
+    return ImmutableSet.of(
+        MIN_NUM_INPUT_FILES,
+        MIN_NUM_OUTPUT_FILES,
+        MIN_FILE_SIZE_BYTES,
+        MAX_FILE_SIZE_BYTES
+    );
+  }
+
+  @Override
+  public RewriteStrategy options(Map<String, String> options) {
+    targetFileSize = PropertyUtil.propertyAsLong(options,
+        RewriteDataFiles.TARGET_FILE_SIZE_BYTES,
+        PropertyUtil.propertyAsLong(
+            table().properties(),
+            TableProperties.WRITE_TARGET_FILE_SIZE_BYTES,
+            TableProperties.WRITE_TARGET_FILE_SIZE_BYTES_DEFAULT));
+
+    minFileSize = PropertyUtil.propertyAsLong(options,
+        MIN_FILE_SIZE_BYTES,
+        (long) (targetFileSize * MIN_FILE_SIZE_DEFAULT_RATIO));
+
+    maxFileSize = PropertyUtil.propertyAsLong(options,
+        MAX_FILE_SIZE_BYTES,
+        (long) (targetFileSize * MAX_FILE_SIZE_DEFAULT_RATIO));
+
+    maxGroupSize = PropertyUtil.propertyAsLong(options,
+        RewriteDataFiles.MAX_FILE_GROUP_SIZE_BYTES,
+        RewriteDataFiles.MAX_FILE_GROUP_SIZE_BYTES_DEFAULT);
+
+    minNumOutputFiles = PropertyUtil.propertyAsInt(options,
+        MIN_NUM_OUTPUT_FILES,
+        MIN_NUM_OUTPUT_FILES_DEFAULT);
+
+    minNumInputFiles = PropertyUtil.propertyAsInt(options,
```

Comment:
These could be validated as well.

Reply:
I think all values are ok here, although anything less than 0 would behave the same as 1, I can prohibit negatives?

Reply:
Makes sense.